Automattic/wp-calypso
# Framework: Add i18n-translate codemod

This codemod:

* Looks for instances of `createClass`-created React components that use `this.translate` (provided by the auto-injected i18n mixin)
* Adds `import { localize } from 'i18n-calypso'` before the first `import` it finds. If there's already something else being `import`ed from `i18n-calypso`, the codemod is smart enough to add the  `{ localize }` import to that statement.
* Wraps the `createClass` instance in `localize()` (thus providing a `translate` prop to the component)
* Replaces `this.translate` with `this.props.translate`

To test:
* `make build`
* `bin/codemods/i18n-mixin client/my-sites/ads/`. Or any directory. Try stuff that doesn't have unit tests (see below).
* (Since this doesn't add whitespace inside `localize`'s parens, do `s/localize\((\S*)\)/localize( $1 )/g`.)

Limitations:
* There needs to be at least one `import` statement at the top of each file you want to change. I'd recommend running @jsnmoon's `./bin/codemods/commonjs-imports path-to-transform/` first -- this will convert `require`s to `import`s.
* If your component has unit tests, the codemod is likely to break them. The solution is to separately export the unwrapped component and use that from inside the unit test; pass lodash's `identity`  there as `translate` prop.
* Wrapping components in `localize` can also break your `refs`.
* Generally speaking, `jscodeshift` based codemods tend to screw up some of our whitespace -- in particular, the spaces inside parens (notably for `localize`).

FYI @jsnmoon @ehg @stephanethomas @gziolo @mtias
